libbpf-tools/lsan.c

```diff
@@ -0,0 +1,1377 @@
+			item->end = end;
+			item->id = begin;
+			if ('\n' == v[MAPS_PATH][0]) {
+				// v[MAPS_PATH] value is "\n" if path name does not exist
+				HASH_FIND(hh, uncertain, &begin, sizeof(uptr), val);
+				if (NULL == val) {
+					item->is_heap = false;
+					HASH_ADD(hh, uncertain, id, sizeof(uptr), item);
+				} else {
+					val->end = end;
+				}
+			} else if (0 == strncmp(heap_str, v[MAPS_PATH], strlen(heap_str))) {
+				HASH_FIND(hh, uncertain, &begin, sizeof(uptr), val);
+				if (NULL == val) {
+					item->is_heap = true;
+					HASH_ADD(hh, uncertain, id, sizeof(uptr), item);
+				} else {
+					val->end = end;
+				}
+			} else if ('[' != v[MAPS_PATH][0] ||
+					0 == strncmp(stack_str, v[MAPS_PATH], strlen(stack_str))) {
+				// not including '[' starting path except "[stack]"
+				HASH_FIND(hh, certain, &begin, sizeof(uptr), val);
+				if (NULL == val) {
+					item->is_heap = false;
+					HASH_ADD(hh, certain, id, sizeof(uptr), item);
+				} else {
+					val->end = end;
+				}
+			}
+		}
+		for (i = 0; i < MAPS_COLUMN_MAX; ++i) {
+			free(v[i]);
+		}
+	}
+	fclose(fp);
+}
+
+static void process_roots(struct root_region_t *roots) {
+	struct root_region_t *curr = NULL;
+	struct root_region_t *next = NULL;
+	HASH_ITER(hh, roots, curr, next) {
+		if (!(curr->is_heap)) {
+			scan_range_for_pointers(curr->id, curr->end, REACHABLE);
+		}
+	}
+}
+
+static void flood_fill_tag(enum chunk_tag tag)
+{
+	struct lsan_hash_t *val = NULL;
+	uptr next_chunk = 0;
+	uptr origin = 0;
+	while (!cvector_empty(frontier)) {
+		next_chunk = frontier[cvector_size(frontier) - 1];
+		cvector_pop_back(frontier);
+		origin = points_into_chunk(next_chunk);
+		HASH_FIND(hh, copied, &origin, sizeof(uptr), val);
+		if (val == NULL)
+			continue;
+
+		scan_range_for_pointers(origin, origin + val->size, tag);
+	}
+}
+
+static pid_t get_tgid(const char *path)
+{
+	__u64 tmp = 0;
+	char buf[BUF_MAX] = {0, };
+	FILE* f = fopen(path, "r");
+	if (!f) {
+		p_warn("Failed to open %s", path);
+		return 0;
+	}
+	while(fscanf(f, "%s %lld\n", buf, &tmp) != EOF) {
+		if (!strcmp(buf, "Tgid:")) {
+			fclose(f);
+			return tmp;
+		}
+	}
+	p_warn("Failed to get tgid: %s\n", path);
+	fclose(f);
+	return 0;
+}
+
+static void get_tids()
+{
+	DIR *folder = opendir("/proc");
+	struct dirent *entry = NULL;
+	char path[PATH_MAX] = {0, };
+	pid_t child_pid = 0;
+	pid_t tgid = 0;
+
+	if (folder == NULL) {
+		p_err("Failed to process registers");
+		p_err("Failed to open /proc directory");
+		return;
+	}
+	entry = readdir(folder);
+	while (entry != NULL) {
+		child_pid = atoi(entry->d_name);
+		if (child_pid != 0) {
+			snprintf(path, sizeof(path), "/proc/%d/status", child_pid);
+			tgid = get_tgid(path);
+			if (tgid == env.tgid) {
+				cvector_push_back(tids, child_pid);
+			}
+		}
+		entry = readdir(folder);
+	}
+	closedir(folder);
+}
+
+static void process_registers()
+{
+	struct user_regs_struct regs;
+	struct iovec io;
+	long ret = 0;
+	int i = 0;
+	int j __attribute__((unused)) = 0;
```

Comment:
I've been thinking about to use `#ifdef` as you said.
But I think it would be better to use `__attribute__((unused))` because other architecture may use this `j` value as well.